### PR TITLE
[#10700] Handle CSRF token problem better

### LINKED ISSUE #10700: Handle CSRF token problem better
Consider cases like the below:
image

Suggestions:
* Make the error message more understandable to users
* Explain workarounds/remedies to the user
* Do not present the submission form to the user (as it is going to fail as well). Instead, give our support email address.

### PR DESCRIPTION
**Pull request by halfwhole**
Fixes #10700

**Outline of Solution**

When the CSRF token is missing, submitting an error form would be inappropriate. This fix detects the particular error on the frontend, changing its error message to the following:

---

### Uh oh! Something went wrong.

We are sorry for the error. It is likely that your browser could not create or access cookies for authorization. If cookies are disabled in your browser, please turn on your cookies before trying again. Alternatively, you could also try emptying your browser's cache, or switching to a different browser.

For further support, please email us at teammates@comp.nus.edu.sg.

---

Would this current message be more appropriate? Reviews and more opinions on this matter would be appreciated.

### CODE REVIEW

File: src/web/app/components/error-report/error-report.component.ts [hunk 1]
```diff
@@ -20,13 +21,19 @@ export class ErrorReportComponent implements OnInit {
...
+  errorReportEnabled: boolean = true;
   errorReportSubmitted: boolean = false;
+  readonly supportEmail: string = environment.supportEmail;
 
   constructor(private errorReportService: ErrorReportService,
               private ngbActiveModal: NgbActiveModal,
               private statusMessageService: StatusMessageService) {}
 
   ngOnInit(): void {
+    if (this.errorMessage === 'Missing CSRF token.') {
```
Comment (ChooJeremy) [inline src/web/app/components/error-report/error-report.component.ts:33]: Are they any other error messages as well that would be related to CSRF token issue? Such as `Invalid CSRF token.`

**Review by ChooJeremy — COMMENTED** [review]
It works, am not a fan of the solution since it requires direct string comparison and an exact match, but I don't see any other easy ways to fix it, since we only return an error message and a HTTP status code.

File: src/web/app/components/error-report/error-report.component.ts (hunk 1 above)
Comment (halfwhole) [reply src/web/app/components/error-report/error-report.component.ts:33]: Yep, `Invalid CSRF token.` is the only other one I could find, I'll update it to include that as well

**Review by ChooJeremy — APPROVED** [review]
LGTM

File: src/web/app/components/error-report/error-report.component.ts [hunk 2]
```diff
@@ -20,13 +21,20 @@ export class ErrorReportComponent implements OnInit {
...
+  csrfErrorMessages: string[] = ['Missing CSRF token.', 'Invalid CSRF token.'];
+  readonly supportEmail: string = environment.supportEmail;
 
   constructor(private errorReportService: ErrorReportService,
               private ngbActiveModal: NgbActiveModal,
               private statusMessageService: StatusMessageService) {}
 
   ngOnInit(): void {
+    if (this.csrfErrorMessages.includes(this.errorMessage)) {
+      this.sendButtonEnabled = false;
```
Comment (madanalogy) [inline src/web/app/components/error-report/error-report.component.ts:35]: Feels a bit redundant since the button wouldn't even be loaded yes?

**Review by madanalogy — CHANGES_REQUESTED** [review]
Am in agreement with Jeremy on this, approach feels dodgy but I can't think of any better way to approach the problem

File: src/web/app/components/error-report/error-report.component.ts (hunk 2 above)
Comment (halfwhole) [reply src/web/app/components/error-report/error-report.component.ts:35]: True, can remove this line

**Comment by halfwhole** [comment]
Perhaps if this approach feels a bit shifty, we could put the error messages as constants in the backend, then make them available to the frontend under `types/api-const`? Helps to standardise the error messages though it still requires direct string matching, so not sure if it is better

**Comment by ChooJeremy** [comment]
> Perhaps if this approach feels a bit shifty, we could put the error messages as constants in the backend, then make them available to the frontend under `types/api-const`? Helps to standardise the error messages though it still requires direct string matching, so not sure if it is better

This would work, but if we're going to try to handle this I would prefer a more proper solution - for example, adding error type and checking based on that. Then perhaps we could have error types based on generic error, authentication error or CSRF token error and process it based on that. Until then, this solution should suffice as a temporary measure.

**Review by madanalogy — APPROVED** [review]
LGTM

**Pull request merged**